src/peers.py

```diff
@@ -130,105 +130,193 @@ def peerline(cls, line):
+    def chrony_peerline(cls, fields):
+        """Convert chrony peer line to named fields"""
         if len(fields) != 10:
-            warn('Invalid peer line - there are %d fields: %s' % (len(fields), line))
+            return None
+        return {
+            'mode': fields[0],
+            'tally': fields[1],
+            'address': fields[2],
+            'stratum': fields[3],
+            'poll_pow2': fields[4],
+            'reach': fields[5],
+            'when': fields[6],
+            'moffset': fields[7],
+            'offset': fields[8],
+            'error': fields[9],
+        }
+
+    ignorerefids = (
+        '.INIT.',
+        '.LOCL.',
+        '.POOL.',
+        '.STEP.',
+        '.XFAC.',
+    )
+
+    @classmethod
+    def ntpd_peerline(cls, tally, fields):
+        """Convert ntpd peer line to named fields"""
+        if len(fields) != 10:
+            return None
+        if fields[1] in cls.ignorerefids:
+            return None
+        return {
+            'tally': tally,
+            'address': fields[0],
+            'refid': fields[1],
+            'stratum': fields[2],
+            'mode': fields[3],
+            'when': fields[4],
+            'poll': fields[5],
+            'reach': fields[6],
+            'delay': fields[7],
+            'offset': fields[8],
+            'jitter': fields[9],
+        }
+
+    @classmethod
+    def validate_tally(cls, fields):
+        peertype = cls.tallytotype(fields['tally'])
+        if peertype == 'unknown':
+            # Unknown peer tally code
             return False
+        else:
+            fields['tally'] = peertype
+        return True
 
-        # stratum should be an integer
+    @classmethod
+    def validate_stratum(cls, fields):
+        # stratum should be an integer between 0 & 15
         try:
-            fields[2] = int(fields[2])
+            fields['stratum'] = int(fields['stratum'])
+            if fields['stratum'] < 0 or fields['stratum'] > 15:
+                # stratum out of bounds
+                return False
         except ValueError:
-            warn('Stratum is not an integer: %s' % fields[2])
+            # stratum not an integer
             return False
+        return True
 
+    @classmethod
+    def validate_when(cls, fields):
         # when should be an integer or '-'
         try:
-            if fields[4] != '-':
-                fields[4] = NTPPeers.time2seconds(fields[4])
+            if fields['when'] != '-':
+                fields['when'] = NTPPeers.time2seconds(fields['when'])
         except ValueError:
-            warn('Last poll time is not an integer: %s' % fields[4])
+            # last poll time not an integer
             return False
+        return True
 
-        # poll should be an integer
+    @classmethod
+    def validate_poll(cls, fields):
+        # poll should be an integer in seconds
+        if 'poll_pow2' in fields:
+            try:
+                fields['poll'] = 2 ** int(fields['poll_pow2'])
```

Comment:
a method named validate_poll shouldn't be resetting the poll value based on some other field; that is an unexpected side effect. It should probably be done when constructing the fields object.

Reply:
Maybe the validate_* methods should be called constrain_* instead?  Their purpose is to ensure that either the values in the fields hash are valid, or that the parse fails.